cli/2024-10-cli-udt.md

```diff
@@ -0,0 +1,150 @@
+- Ensure that the CLI experience is consistent with the other commands in Radius
+
+### Non-goals (out of scope)
+<!-- What are we explicitly not trying to accomplish? -->
+- Other interfaces and experiences to interact with UDTs are out of scope for this document
+
+## Key scenarios
+<!-- List ~3-7 high level scenarios to clarify the value and point to how we will decompose this big area into component capabilities. We may ultimately have more than one level of scenario. -->
+The key scenarios for UDTs are outlined [here](/architecture/2024-06-resource-extensibility-feature-spec.md). We will decompose these scenarios specific to the CLI experience.
+
+### Scenario 1: Deb defines and deploys the internal messaging service using the CLI
+<!-- One or two sentence summary -->
+Deb, a platform engineer who wants to define and deploy their internal messaging service Plaid as a resource type in Radius. 
+
+Before we start designing the experience, we need to understand the high level terminologies/concept that a user would interact with in the CLI experience.
+
+| Terminology | Description | Example |
+|-------------|-------------|---------|
+|Resource provider | Refers to a group of resource types under a common namespace. Resource providers are also entities that implement the group of resource types | Mycompany.Messaging or Applications.Core |
+|Resource type | A service or technology that can be modeled as a resource | PostgreSQL or AWS S3 or Internal messaging service |
+|Resource | An instance of a resource type | mydb or mybucket |
+
+#### User Experience
+
+1. Deb uses the rad CLI to scaffold a `yaml` template definition for the internal messaging service `Plaid`
+
+    Option 1:
+
+    ```bash
+    rad resource-provider init Mycompany.Messaging --schema-format yaml
+    ```
+        
+    This command will create a folder with the name `Mycompany.Messaging`. A sample `yaml` file is created in the application folder. If we enable typespec, a sample `tsp` will be created and all the required dependencies will be created.
+
+    Pros : Depicts the action of initializing the project with directory structure and dependencies. TypeSpec has similar `init` command to set up the project and would be consistent with that when we support typespec.
+
+    Option 2:
+
+    ```bash
+    rad resource-type new Mycompany.Messaging --schema-format yaml
+    ```
+    
+    Deb updates the schema definition with the required fields and properties for the messaging service Plaid. The authoring experience for the resource type will be covered in a separate document.
+
+2. Deb registers the resource type with Radius
+
+    Option 1 
+
+    ```bash
+    rad resource-provider create Mycompany.Messaging
+    ```
+    Pros : Follows the CRUDL convention and is more consistent with the other commands in Radius
+
+    Option 2
+
+    ```bash
+    rad resource-provider register Mycompany.Messaging
```

Comment:
I'd love for us to have consistent verbs. I would propose that `create` is the verb for creating new objects within Radius. `register` is for linking to existing things, like registering a cloud provider. It's unfortunate that `rad recipe register` is done that way since it's inconsistent with this definition.

Reply:
That helps me understand the thinking on recipe register. Thanks.